<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Mapping QTLs in outbred mice using varbvs • varbvs</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script><script src="../pkgdown.js"></script><!-- mathjax --><script src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-vignette">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">varbvs</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../articles/index.html">Articles</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="http://github.com/pcarbo/varbvs">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9">
    <div class="page-header toc-ignore">
      <h1>Mapping QTLs in outbred mice using varbvs</h1>
            
            <h4 class="date">2017-04-04</h4>
          </div>

    
    
<div class="contents">
<p>In this vignette, we use <strong>varbvs</strong> to map QTLs for phenotypes measured in CFW (Carworth Farms White) outbred mice. Phenotypes include muscle weights—EDL and soleus muscle—and testis weight measured at sacrifice. Running this script with <code>trait = "testis"</code> reproduces the results and figures shown in the second example of a forthcoming paper (Carbonetto <em>et al</em>, 2016).</p>
<div id="vignette-parameters" class="section level2">
<h2 class="hasAnchor">
<a href="#vignette-parameters" class="anchor"></a>Vignette parameters</h2>
<p>Begin by loading packages into the R environment.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(lattice)
<span class="kw">library</span>(varbvs)</code></pre></div>
<p>These script parameters specify the candidate prior log-odds settings, the prior variance of the coefficients, and which trait to analyze. Set trait to “edl”, “soleus” or “testis”.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">trait   &lt;-<span class="st"> "testis"</span>
logodds &lt;-<span class="st"> </span><span class="kw">seq</span>(-<span class="dv">5</span>,-<span class="dv">3</span>,<span class="fl">0.25</span>)
sa      &lt;-<span class="st"> </span><span class="fl">0.05</span></code></pre></div>
<p>Set the random number generator seed.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">set.seed</span>(<span class="dv">1</span>)</code></pre></div>
</div>
<div id="load-the-genotype-and-phenotype-data" class="section level2">
<h2 class="hasAnchor">
<a href="#load-the-genotype-and-phenotype-data" class="anchor"></a>Load the genotype and phenotype data</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">load</span>(<span class="st">"cfw.RData"</span>)
if (trait ==<span class="st"> "edl"</span> |<span class="st"> </span>trait ==<span class="st"> "soleus"</span>) {
  covariates &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">"batch16"</span>,<span class="st">"tibia"</span>)
} else if (trait ==<span class="st"> "testis"</span>) {
  covariates &lt;-<span class="st"> "sacwt"</span>
}</code></pre></div>
<p>Only analyze samples for which the phenotype and all the covariates are observed.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">rows &lt;-<span class="st"> </span><span class="kw">which</span>(<span class="kw">apply</span>(pheno[,<span class="kw">c</span>(trait,covariates)],<span class="dv">1</span>,
                    function (x) <span class="kw">sum</span>(<span class="kw">is.na</span>(x)) ==<span class="st"> </span><span class="dv">0</span>))
pheno &lt;-<span class="st"> </span>pheno[rows,]
geno  &lt;-<span class="st"> </span>geno[rows,]</code></pre></div>
</div>
<div id="fit-variational-approximation-to-posterior" class="section level2">
<h2 class="hasAnchor">
<a href="#fit-variational-approximation-to-posterior" class="anchor"></a>Fit variational approximation to posterior</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">runtime &lt;-<span class="st"> </span><span class="kw">system.time</span>(fit &lt;-
<span class="st">  </span><span class="kw"><a href="../reference/varbvs.html">varbvs</a></span>(geno,<span class="kw">as.matrix</span>(pheno[,covariates]),pheno[,trait],
         <span class="dt">sa =</span> sa,<span class="dt">logodds =</span> logodds,<span class="dt">verbose =</span> <span class="ot">FALSE</span>))
<span class="kw">cat</span>(<span class="kw">sprintf</span>(<span class="st">"Model fitting took %0.2f minutes.</span><span class="ch">\n</span><span class="st">"</span>,runtime[<span class="st">"elapsed"</span>]/<span class="dv">60</span>))</code></pre></div>
</div>
<div id="summarize-the-results-of-model-fitting" class="section level2">
<h2 class="hasAnchor">
<a href="#summarize-the-results-of-model-fitting" class="anchor"></a>Summarize the results of model fitting</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">print</span>(<span class="kw">summary</span>(fit))</code></pre></div>
<p>Show three genome-wide scans: (1) one using the posterior inclusion probabilities (PIPs) computed in the BVS analysis of all SNPs; (2) one using the p-values computed using GEMMA; and (3) one using the PIPs computed from the BVSR model in GEMMA.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">trellis.par.set</span>(<span class="dt">axis.text     =</span> <span class="kw">list</span>(<span class="dt">cex =</span> <span class="fl">0.7</span>),
                <span class="dt">par.ylab.text =</span> <span class="kw">list</span>(<span class="dt">cex =</span> <span class="fl">0.7</span>),
                <span class="dt">par.main.text =</span> <span class="kw">list</span>(<span class="dt">cex =</span> <span class="fl">0.7</span>,<span class="dt">font =</span> <span class="dv">1</span>))
w &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="kw"><a href="../reference/normalizelogweights.html">normalizelogweights</a></span>(fit$logw))
j &lt;-<span class="st"> </span><span class="kw">which</span>(fit$alpha %*%<span class="st"> </span>w &gt;<span class="st"> </span><span class="fl">0.5</span>)
r &lt;-<span class="st"> </span>gwscan.gemma[[trait]]
r[<span class="kw">is.na</span>(r)] &lt;-<span class="st"> </span><span class="dv">0</span>
chromosomes   &lt;-<span class="st"> </span><span class="kw">levels</span>(gwscan.bvsr$chr)
xticks        &lt;-<span class="st"> </span><span class="kw">rep</span>(<span class="dv">0</span>,<span class="kw">length</span>(chromosomes))
<span class="kw">names</span>(xticks) &lt;-<span class="st"> </span>chromosomes
pos           &lt;-<span class="st"> </span><span class="dv">0</span>
for (i in chromosomes) {
  n         &lt;-<span class="st"> </span><span class="kw">sum</span>(gwscan.bvsr$chr ==<span class="st"> </span>i)
  xticks[i] &lt;-<span class="st"> </span>pos +<span class="st"> </span>n/<span class="dv">2</span>
  pos       &lt;-<span class="st"> </span>pos +<span class="st"> </span>n +<span class="st"> </span><span class="dv">25</span>
}
<span class="kw">print</span>(<span class="kw">plot</span>(fit,<span class="dt">groups =</span> map$chr,<span class="dt">vars =</span> j,<span class="dt">gap =</span> <span class="dv">1500</span>,<span class="dt">cex =</span> <span class="fl">0.6</span>,
           <span class="dt">ylab =</span> <span class="st">"probability"</span>,<span class="dt">main =</span> <span class="st">"a. multi-marker (varbvs)"</span>,
           <span class="dt">scales =</span> <span class="kw">list</span>(<span class="dt">y =</span> <span class="kw">list</span>(<span class="dt">limits =</span> <span class="kw">c</span>(-<span class="fl">0.1</span>,<span class="fl">1.2</span>),<span class="dt">at =</span> <span class="kw">c</span>(<span class="dv">0</span>,<span class="fl">0.5</span>,<span class="dv">1</span>))),
           <span class="dt">vars.xyplot.args =</span> <span class="kw">list</span>(<span class="dt">cex =</span> <span class="fl">0.6</span>)),
      <span class="dt">split =</span> <span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">1</span>,<span class="dv">1</span>,<span class="dv">3</span>),<span class="dt">more =</span> <span class="ot">TRUE</span>)
<span class="kw">print</span>(<span class="kw">plot</span>(fit,<span class="dt">groups =</span> map$chr,<span class="dt">score =</span> r,<span class="dt">vars =</span> j,<span class="dt">cex =</span> <span class="fl">0.6</span>,<span class="dt">gap =</span> <span class="dv">1500</span>,
           <span class="dt">draw.threshold =</span> <span class="fl">5.71</span>,<span class="dt">ylab =</span> <span class="st">"-log10 p-value"</span>,
           <span class="dt">main =</span> <span class="st">"b. single-marker (GEMMA -lm 2)"</span>,
           <span class="dt">scales =</span> <span class="kw">list</span>(<span class="dt">y =</span> <span class="kw">list</span>(<span class="dt">limits =</span> <span class="kw">c</span>(-<span class="dv">1</span>,<span class="dv">20</span>),<span class="dt">at =</span> <span class="kw">seq</span>(<span class="dv">0</span>,<span class="dv">20</span>,<span class="dv">5</span>))),
           <span class="dt">vars.xyplot.args =</span> <span class="kw">list</span>(<span class="dt">cex =</span> <span class="fl">0.6</span>)),
     <span class="dt">split =</span> <span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">1</span>,<span class="dv">3</span>),<span class="dt">more =</span> <span class="ot">TRUE</span>)
<span class="kw">print</span>(<span class="kw">xyplot</span>(p1 ~<span class="st"> </span>plot.x,gwscan.bvsr,<span class="dt">pch =</span> <span class="dv">20</span>,<span class="dt">col =</span> <span class="st">"midnightblue"</span>,
             <span class="dt">scales =</span> <span class="kw">list</span>(<span class="dt">x =</span> <span class="kw">list</span>(<span class="dt">at =</span> xticks,<span class="dt">labels =</span> chromosomes),
                           <span class="dt">y =</span> <span class="kw">list</span>(<span class="dt">limits =</span> <span class="kw">c</span>(-<span class="fl">0.1</span>,<span class="fl">1.2</span>),<span class="dt">at =</span> <span class="kw">c</span>(<span class="dv">0</span>,<span class="fl">0.5</span>,<span class="dv">1</span>))),
             <span class="dt">xlab =</span> <span class="st">""</span>,<span class="dt">ylab =</span> <span class="st">"probability"</span>,<span class="dt">main =</span> <span class="st">"c. multi-marker (BVSR)"</span>),
      <span class="dt">split =</span> <span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">3</span>,<span class="dv">1</span>,<span class="dv">3</span>),<span class="dt">more =</span> <span class="ot">FALSE</span>)</code></pre></div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#vignette-parameters">Vignette parameters</a></li>
      <li><a href="#load-the-genotype-and-phenotype-data">Load the genotype and phenotype data</a></li>
      <li><a href="#fit-variational-approximation-to-posterior">Fit variational approximation to posterior</a></li>
      <li><a href="#summarize-the-results-of-model-fitting">Summarize the results of model fitting</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Peter Carbonetto, Matthew Stephens, David Gerard.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  </body>
</html>
